File: day1/day1-2.py
digit_words = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
digit_length_3 = ["one", "two", "six"]
digit_length_4 = ["four", "five", "nine"]
digit_length_5 = ["three", "seven", "eight"]


def word_to_digit(word: str) -> int:
    return digit_words.index(word) + 1


def get_first_digit(string: list[str]) -> int:
    length = len(string)

    for i in range(length):
        if string[i].isdecimal():
            return int(string[i])

        words_3 = "".join(string[i : i + 3])
        words_4 = "".join(string[i : i + 4])
        words_5 = "".join(string[i : i + 5])

        if i <= length - 3 and words_3 in digit_length_3:
            return word_to_digit("".join(words_3))
        elif i <= length - 4 and words_4 in digit_length_4:
            return word_to_digit("".join(words_4))
        elif i <= length - 5 and words_5 in digit_length_5:
            return word_to_digit("".join(words_5))


def get_second_digit(string: list[str]) -> int:
    length = len(string)

    for i in reversed(range(length)):
        if string[i].isdecimal():
            return int(string[i])

        words_3 = "".join(string[i - 2 : i + 1])
        words_4 = "".join(string[i - 3 : i + 1])
        words_5 = "".join(string[i - 4 : i + 1])

        if i >= 2 and words_3 in digit_length_3:
            return word_to_digit("".join(words_3))
        elif i >= 3 and words_4 in digit_length_4:
            return word_to_digit("".join(words_4))
        elif i >= 4 and words_5 in digit_length_5:
            return word_to_digit("".join(words_5))


def get_calibration_value(string: list[str]):
    calibration_value_list = list()

    calibration_value_list.append(get_first_digit(string))
    calibration_value_list.append(get_second_digit(string))

    return calibration_value_list[0] * 10 + calibration_value_list[1]


with open("day1-2.input", "r") as f:
    data = list(map(list, f.readlines()))

    print(sum(map(get_calibration_value, data)))

File: kotlin/src/main/kotlin/day1/Day1_2.kt
package day1

val digit_words = listOf(
    "one", "two", "three", "four",
    "five", "six", "seven", "eight", "nine"
)
val digit_length_3 = listOf("one", "two", "six")
val digit_length_4 = listOf("four", "five", "nine")
val digit_length_5 = listOf("three", "seven", "eight")

fun wordToDigit(word: String): Int {
    return digit_words.indexOf(word) + 1
}

fun getFirstDigit(input: String): Int {
    for ((index, char) in input.withIndex()) {
        if (char.isDigit()) {
            return Integer.parseInt(char.toString())
        }

        if (index + 3 < input.length) {
            val words3 = input.substring(index, index + 3)
            if (words3 in digit_length_3) {
                return wordToDigit(words3)
            }
        }

        if (index + 4 < input.length) {
            val words4 = input.substring(index, index + 4)
            if (words4 in digit_length_4) {
                return wordToDigit(words4)
            }
        }

        if (index + 5 < input.length) {
            val words5 = input.substring(index, index + 5)
            if (words5 in digit_length_5) {
                return wordToDigit(words5)
            }
        }
    }
    throw Exception("No digit found")
}

fun getLastDigit(input: String): Int {
    for ((index, char) in input.reversed().withIndex()) {
        if (char.isDigit()) {
            return Integer.parseInt(char.toString())
        }

        if (input.length - index - 3 >= 0) {
            val words3 = input.substring(input.length - index - 3, input.length - index)
            if (words3 in digit_length_3) {
                return wordToDigit(words3)
            }
        }

        if (input.length - index - 4 >= 0) {
            val words4 = input.substring(input.length - index - 4, input.length - index)
            if (words4 in digit_length_4) {
                return wordToDigit(words4)
            }
        }

        if (input.length - index - 5 >= 0) {
            val words5 = input.substring(input.length - index - 5, input.length - index)
            if (words5 in digit_length_5) {
                return wordToDigit(words5)
            }
        }
    }
    throw Exception("No digit found")
}

fun getCalibrationValue2(input: String): Int {
    return getFirstDigit(input) * 10 + getLastDigit(input)
}

fun main() {
    val inputs = readFile()
    println(inputs.sumOf { getCalibrationValue2(it) })
}
